src/inference/tests/unit/memory_solver_test.cpp

```diff
@@ -0,0 +1,286 @@
+    };
+
+    MemorySolver ms(boxes);
+    EXPECT_EQ(ms.solve(), 6);
+    EXPECT_EQ(ms.maxDepth(), 6);
+    EXPECT_EQ(ms.maxTopDepth(), 2);
+}
+
+TEST(MemSolverTest, EndOnSeveralBegins) {
+    std::vector<Box> boxes{
+        //  |      ____
+        {0, 1, 2},  //  |     |____| ____
+        {1, 2, 2},  //  |           |____|__
+        {3, 3, 2},  //  |   ____    |_______|
+        {3, 5, 2},  //  |__|____|___|_|_________
+        {3, 4, 2},  //      0  1  2  3  4  5  6
+    };
+
+    MemorySolver ms(boxes);
+    EXPECT_EQ(ms.solve(), 6);
+    EXPECT_EQ(ms.maxDepth(), 6);
+    EXPECT_EQ(ms.maxTopDepth(), 3);
+}
+
+TEST(MemSolverTest, ToEndBoxes) {
+    std::vector<Box> boxes{
+        //  |      _____________
+        {0, 1, 2},   //  |     |_____________>>
+        {1, -1, 2},  //  |           |____|__
+        {3, 3, 2},   //  |   ____    |_______>>
+        {3, -1, 2},  //  |__|____|___|_|_________
+        {3, 4, 2},   //      0  1  2  3  4  5  6
+    };
+
+    MemorySolver ms(boxes);
+    EXPECT_EQ(ms.solve(), 8);
+    EXPECT_EQ(ms.maxDepth(), 8);
+    EXPECT_EQ(ms.maxTopDepth(), 4);
+}
+
+TEST(MemSolverTest, LastAndToEndBox) {
+    std::vector<Box> boxes{
+        //  |                     _
+        {0, 1, 2},   //  |            ____    |_>>
+        {6, -1, 2},  //  |           |____|__
+        {3, 3, 2},   //  |   ____    |_______|
+        {3, 5, 2},   //  |__|____|___|_|_________
+        {3, 4, 2},   //      0  1  2  3  4  5  6
+    };
+
+    MemorySolver ms(boxes);
+    EXPECT_EQ(ms.solve(), 6);
+    EXPECT_EQ(ms.maxDepth(), 6);
+    EXPECT_EQ(ms.maxTopDepth(), 3);
+}
+
+TEST(MemSolverTest, OptimalAlexnet) {
+    std::vector<std::vector<int>> shapes{
+        {3, 227, 227},  // in
+        {96, 55, 55},   // conv1
+        {96, 55, 55},   // relu1
+        {96, 55, 55},   // norm1
+        {96, 27, 27},   // pool1
+        {256, 27, 27},  // conv2
+        {256, 27, 27},  // relu2
+        {256, 27, 27},  // norm2
+        {256, 13, 13},  // pool2
+        {384, 13, 13},  // conv3
+        {384, 13, 13},  // relu3
+        {384, 13, 13},  // conv4
+        {384, 13, 13},  // relu4
+        {256, 13, 13},  // conv5
+        {256, 13, 13},  // relu5
+        {256, 6, 6},    // pool5
+        {1, 1, 4069},   // fc6
+        {1, 1, 4069},   // relu6
+        {1, 1, 4069},   // fc7
+        {1, 1, 4069},   // relu7
+        {1, 1, 1000},   // fc8
+        {1, 1, 1000},   // loss
+    };
+
+    int n = 0;
+    std::vector<Box> boxes;
+    for (const auto& sh : shapes)
+        boxes.push_back({n, ++n, sh[0] * sh[1] * sh[2]});
+
+    // For linear topology bottom score is reachable minRequired == maxDepth
+    MemorySolver ms(boxes);
+    EXPECT_EQ(ms.solve(), ms.maxDepth());
+    EXPECT_EQ(ms.maxTopDepth(), 2);
+}
+
+TEST(MemSolverTest, NoOverlapping) {
+    int n = 0;  //  |         _____________
+    std::vector<Box> boxes{
+        //  |   _____|___1_________|
+        {4, 8, 1, n++},  //  |  |_2_____|    ____
```

Comment:
Clang format breaks comments. Please double check (and for all other PRs with tests movement.
Maybe we can enable clang format separatelly